### DISCUSSION TITLE
Extracting meta data from YAML using Quarto Shortcodes

### DISCUSSION THREAD

**Discussion by produnis** [quarto-dev/quarto-cli#8241] [Q&A] [answered]
### Description

Hello everyone,

I'm currently learning how to create shortcodes, and I'm encountering difficulties extracting meta-data from YAML into my code.

As an example, I'm trying to retrieve the ORCID number from the author field. The goal is to have `{{< orcid-entry >}}` render as the link` <a href="...">Orcid</a>.`

Here's an example of my YAML:

```
---
author:
  - name: Hans Wurst
    orcid: 0000-0002-xyza-foobar
---
```

And this is how my Lua filter looks:

```

return {
  ['orcid-entry'] = function(meta)
    if meta.author then
      for _, author in ipairs(meta.author) do
        if author.orcid then
          local orcid_link = "<a href=\"https://orcid.org/" .. author.orcid .. "\">orcid</a>"
          return pandoc.Para(pandoc.RawInline("html", orcid_link))
        end
      end
    end
    return pandoc.Para(pandoc.Str("ORCID not found"))
  end
}

```

My problem is that I always get "ORCID not found," even when ORCID is set in YAML. What am I doing wrong?

**Comment by cderv** [comment]
Unfortunately this is really badly documented on our side - really sorry about this. We'll improve ! 

`author` metadata is a special one that get normalized by Quarto to improve user experience in providing metadata and facilitate the development in Lua and in custom templates. This is detailed in https://quarto.org/docs/journals/authors.html#author-metadata

Basically, `author` fields will be split in `authors` and `by-authors` so the value you want will be in those fields. A trick so see what is happening is to render as native `quarto render index.qmd --to native` you test file. 

So 

* you need to look in `authors` or `by-authors`
* and you probably need `pandoc.utils.stringify()` on your orcid value before concatenating with other strings. 

Hope it helps

**Reply by mcanouil** [reply]
FYI, for convenience using "Preview", one can write in the YAML frontmatter `format: native`.
Also, Lua logging can help when developing with Lua: https://quarto.org/docs/extensions/lua.html#diagnostic-logging

**Reply by produnis** [reply, marked as answer]
Fortunately, I got the answer [here](https://stackoverflow.com/questions/77882406/how-to-access-meta-data-from-quarto-yaml-in-lua-script/77884499#77884499). In addition to your suggestions, my script was lacking the right function at the beginning.. ;) As it is `function(args, kwargs, meta)`, the data is in `args`, not in `meta`...
The working lua script looks like this now:

```
local str = pandoc.utils.stringify

return {
  ['orcid'] = function(args, kwargs, meta)
    local author_id = str(args[1])
    local authors = meta.authors or meta['by-author']

    if authors then
      for _, author in ipairs(authors) do
        if author.id == author_id then
          if author.orcid then
            local orcid_value = str(author.orcid)
            local orcid_link = "<a href=\"https://orcid.org/" .. orcid_value .. "\">orcid</a>"
            return pandoc.Para(pandoc.RawInline("html", orcid_link))
          end
          return pandoc.Para(pandoc.Str("ORCID not found"))
        end
      end
    end

  end
}
```

and you trigger it with `{{< orcid 1 >}}`, `{{< orcid 2 >}}` ...
thx for your help!
